### ISSUE TITLE
hresult_error doesn't play well with C++ unit test frameworks

### ISSUE THREAD

**Issue by davidmatson** [microsoft/cppwinrt#1129]
Some additional details related to #207. Perhaps would need to be v3/#1123.

Consider the following test code:
```c++
#include "gtest/gtest.h"
#include <string_view>
#include <winrt/base.h>
#include <windows.h>

struct non_std_exception
{
	explicit non_std_exception(std::string_view what) : m_what{ what }
	{
	}

	std::string_view what()
	{
		return m_what;
	}
private:
	std::string m_what;
};

inline HRESULT do_something()
{
	try
	{
		throw std::runtime_error{ "a bad thing happened" };
	}
	catch (...)
	{
		return winrt::to_hresult();
	}
}

TEST(SampleFixture, SampleTest)
{
	//throw std::runtime_error{ "a bad thing happened" };
	//throw non_std_exception{ "a bad thing happened" };
	winrt::check_hresult(do_something());
}
```

which basically shows what happens when the implementation of a cppwinrt component throws an exception that's during a unit test, in particular, any std::exception-derived exception that has a message/.what().

Here's the output:
```
[ RUN      ] SampleFixture.SampleTest
unknown file: error: Unknown C++ exception thrown in the test body.
[  FAILED  ] SampleFixture.SampleTest (5 ms)
```

When uncommenting the std::runtime_error line instead, here's the output:
```
[ RUN      ] SampleFixture.SampleTest
unknown file: error: C++ exception with description "a bad thing happened" thrown in the test body.
[  FAILED  ] SampleFixture.SampleTest (0 ms)
[----------] 1 test from SampleFixture (1 ms total)
```

The sample above uses GoogleTest, but the same holds with Boost.Test and Catch2. The universal contract between a unit test and a c++ unit testing framework appears to be:
The test passes unless an exception is thrown.
If an exception is thrown, the message from std::exception.what() is displayed in the test failure message.

Furthermore, test failure messages are not customizable from any of the c++ unit test frameworks I checked - they apparently follow the [Standard C++ guidance](https://isocpp.org/wiki/faq/exceptions#what-to-throw) that exceptions should (ultimately) derive from std::exception, if possible.
(A couple also support directly throwing a char* or std::string and providing that text in the test failure message; none I checked support custom exception types.)

I understand there would be a UTF-8 conversion cost to pay, but for unit test failures, the experience is basically undebuggable when the exception message gets dropped and all you get is "unknown C++ exception".

Two main questions:
1. Could something better be done here, especially as part of v3 planning?
2. Are there any hooks/workarounds you'd recommend in the meantime that would make exception messages "just work" in unit test failures?

Thanks!

**Comment by sylveon** [comment]
Not deriving from std::exception is a design decision that dates from back when C++/WinRT was first designed (as you've pointed out by linking #207). I'm not sure deriving from std::exception is possible at this point (since it can lead to silent source breakage - for example code catching (in this order) both std::exception and winrt::hresult_error will still compile, but the latter will not run anymore) and the concerns that lead to the decision still apply, AFAIK.

On the flipside, I'm surprised these unit testing frameworks don't support customizing output for custom exceptions, as C++/WinRT is certainly not the only one using custom exceptions

**Comment by kennykerr** [comment]
Yes, testing frameworks should typically provide some way to hook exceptions as we do here:

https://github.com/microsoft/cppwinrt/blob/b69f40d6ecc7d852aebb0682e099322b60cc34e3/test/test/main.cpp#L14-L17

**Comment by davidmatson** [comment]
> Yes, testing frameworks should typically provide some way to hook exceptions as we do here:
> 
> https://github.com/microsoft/cppwinrt/blob/b69f40d6ecc7d852aebb0682e099322b60cc34e3/test/test/main.cpp#L14-L17

Unfortunately, it doesn't seem like any of them do. I debugged through MSTest, GoogleTest, Boost.Test, and Catch2 and couldn't find a way to hook exceptions in any of them.

My guess is they're following the isocpp guidance that says "derive (ultimately) std::exception" as their basis for keeping things simpler that way.

That fact that so many of them say this is the contract, and isocpp gives the same guideline, feels pretty compelling for me.

**Comment by davidmatson** [comment]
Yeah, sorry, I just made that connection. I wonder if it's partly a doc issue:
https://catch2.docsforge.com/v2.13.2/macros-api/CATCH_TRANSLATE_EXCEPTION/
(There isn't much there.)

Revisiting for cppwinrt v3 sounds great - thanks!

**Comment by davidmatson** [comment]
It looks like GoogleTest doesn't currently provide such a mechanism; there's an open request:
https://github.com/google/googletest/issues/3433

So for v3, I think there would be value in reconsidering.
